packages/3d-web-voice-chat/src/voice-chat-manager/VoiceChatManager.ts

```diff
@@ -0,0 +1,331 @@
+import { CharacterState } from "@mml-io/3d-web-client-core/src";
+import VoxeetSDK from "@voxeet/voxeet-web-sdk";
+import Conference from "@voxeet/voxeet-web-sdk/types/models/Conference";
+import type { SpatialAudioStyle } from "@voxeet/voxeet-web-sdk/types/models/SpatialAudio";
+
+import MicrophoneOff from "../chat-ui/icons/MicrophoneOff.svg";
+import MicrophoneOn from "../chat-ui/icons/MicrophoneOn.svg";
+
+import { getEulerFromQuaternion, formatPos, formatDirection } from "./helpers";
+
+export type Quaternion = { w: number; x: number; y: number; z: number };
+
+export type Euler = { pitch: number; yaw: number; roll: number };
+
+export type Position = { x: number; y: number; z: number };
+
+export type Rotation = { quaternionY: number; quaternionW: number };
+
+type Direction = { x: number; y: number; z: number };
+
+type Vector3 = {
+  x: number;
+  y: number;
+  z: number;
+};
+
+type ConferenceSettings = {
+  scale: Vector3;
+  forward: Vector3;
+  up: Vector3;
+  right: Vector3;
+};
+
+export class VoiceChatManager {
+  private debug = false;
+
+  private disposed = false;
+  private pending = false;
+  private hasJoinedAudio = false;
+  private speaking = false;
+
+  private conferenceAlias: string;
+  private participants = new Map<string, string>();
+  private activeSpeakers: number = 0;
+
+  private micOnIcon: string = `<img src="data:image/svg+xml;utf8,${encodeURIComponent(
+    MicrophoneOn,
+  )}" />`;
+  private micOffIcon: string = `<img src="data:image/svg+xml;utf8,${encodeURIComponent(
+    MicrophoneOff,
+  )}" />`;
+
+  private conference: Conference | null = null;
+
+  private settings: ConferenceSettings = {
+    scale: { x: 2, y: 2, z: 2 },
+    forward: { x: 0, y: -1, z: 0 },
+    up: { x: 0, y: 0, z: 1 },
+    right: { x: -1, y: 0, z: 0 },
+  };
+
+  private joinButton: HTMLElement | null = null;
+  private participantsDiv: HTMLElement | null = null;
+
+  private tickInterval: NodeJS.Timeout | null = null;
+
+  constructor(
+    private userId: number,
+    private remoteUserStates: Map<number, CharacterState>,
+    private latestCharacterObj: {
+      characterState: null | CharacterState;
+    },
+  ) {
+    this.conferenceAlias = window.location.host;
+    this.init();
+    this.tickInterval = setInterval(() => this.tick(), 1000);
+  }
+
+  private setStatusElement() {
+    if (!this.joinButton) {
+      this.joinButton = document.getElementById("voice-join-button");
```

Comment:
This binding between the UI and the manager feels brittle and difficult to follow. The manager should probably be given a way to interface with the UI directly.

Reply:
`VoiceChatManager` and `voice-chat-ui` completely refactored in [this commit](https://github.com/mml-io/3d-web-experience/pull/42/commits/31c10d5991711251cf1907fc19c301a442114fdb) so the manager gets responsible for (re)rendering the UI through props.